overlord/configstate/configcore/netplan.go

```diff
@@ -92,18 +117,137 @@ func getNetplanFromSystem(key string) (result interface{}, err error) {
 	netplan := conn.Object("io.netplan.Netplan", "/io/netplan/Netplan")
 
 	if err := netplan.Call("io.netplan.Netplan.Config", 0).Store(&netplanConfigSnapshotBusAddr); err != nil {
-		if derr, ok := err.(dbus.Error); ok {
-			// Having no netplan config is *not* an error, we just
-			// do not support netplan config.
-			if isNoServiceOrMethodErr(derr) {
-				return nil, nil
-			}
+		return nil, err
+	}
+	logger.Debugf("using netplan config %v", netplanConfigSnapshotBusAddr)
+
+	netplanCfgSnapshot := conn.Object("io.netplan.Netplan", dbus.ObjectPath(netplanConfigSnapshotBusAddr))
+	return netplanCfgSnapshot, nil
+}
+
+func validateNetplanSettings(tr config.Conf) error {
+	// XXX: validate somehow once we support writing?
+	return nil
+}
+
+func hasNetplanChanges(tr config.Conf) bool {
+	for _, chg := range tr.Changes() {
+		if chg == "core.system.network.netplan" || strings.HasPrefix(chg, "core.system.network.netplan.") {
+			return true
+		}
+	}
+	return false
+}
+
+func handleNetplanConfiguration(tr config.Conf, opts *fsOnlyContext) error {
+	if !hasNetplanChanges(tr) {
+		return nil
+	}
+
+	var cfg map[string]interface{}
+	if err := tr.Get("core", "system.network.netplan", &cfg); err != nil && !config.IsNoOption(err) {
+		return err
+	}
+
+	netplanCfgSnapshot, err := getNetplanCfgSnapshot()
+	// Having no netplan config is *not* an error, we just
+	// do not support netplan config.
+	if isNoServiceOrMethodErr(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Use the default config that most Ubuntu Core have
+	originHint := "90-snapd-config"
+
+	// Always starts with a clean config to avoid merging of keys
+	// that got unset.
+	configs := []string{"network=null"}
+	// and then pass the full new config in
+	for key := range cfg {
+		// We pass the new config back to netplan as json, the reason
+		// is that the dbus api accepts only a single line string, see
+		// see https://github.com/canonical/netplan/pull/210
+		jsonNetplanConfigRaw, err := json.Marshal(cfg[key])
+		if err != nil {
+			return fmt.Errorf("cannot netplan config: %v", err)
+		}
+		configs = append(configs, fmt.Sprintf("%s=%s", key, string(jsonNetplanConfigRaw)))
+	}
+
+	// now apply
+	for _, jsonNetplanConfig := range configs {
+		logger.Debugf("calling netplan.Set: %v", jsonNetplanConfig)
+
+		var wasSet bool
+		if err := netplanCfgSnapshot.Call("io.netplan.Netplan.Config.Set", 0, jsonNetplanConfig, originHint).Store(&wasSet); err != nil {
+			return fmt.Errorf("cannot Set() config: %v", err)
+		}
+	}
+
+	var storeReachableBefore bool
+	if err := storeReachable(tr.State()); err == nil {
+		storeReachableBefore = true
+	}
+	logger.Debugf("store reachable before netplan %v", storeReachableBefore)
+
+	var wasTried bool
+	timeoutInSeconds := 30
+	if err := netplanCfgSnapshot.Call("io.netplan.Netplan.Config.Try", 0, uint32(timeoutInSeconds)).Store(&wasTried); err != nil {
```

Comment:
Do we know that when Try returns, the configuration has been applied? What if we have a wifi uplink and it takes time to connect to a configured AP, does Try also wait for the connection to happen before giving up?

On a side note, can this timeout in go-dbus if Try waits?

Reply:
This is an excellent point, I don't know the exact semantics of "try", maybe @slyon can help out here? In our connectivity checks we use `httputil.RetryRequest()` so we do a certain level of retrying but maybe it's actually best to put a look around this one to ensure we retry for a couple of seconds?

Reply:
Essentially, what netplan does is calling `networkctl reload` && `netplanctl reconfigure IFACE` (or restarting NetworkManager.service accordingly).

So it waits for the interfaces to be configured and acknowledged by the underlying networking daemon (networkd/NetworkManager), but does not wait for the network being online/connected. DHCP response or connecting to an AP might not necessarily be done by this point.

There are other tools that can be  used for this job, like `/usr/lib/systemd/systemd-networkd-wait-online`, but there is another problem with this: We do not have a common understanding/definition of what an "online" systems is. (That is being worked on by Foundations this cycle in FR-853)

Reply:
Thanks, I added a retry now